internal/controller/issues.go

```diff
@@ -0,0 +1,180 @@
+	if githubIssue == nil {
+		return false
+	}
+	state := githubIssue.GetState()
+	condition := &metav1.Condition{
+		Type:    "IssueIsOpen",
+		Status:  metav1.ConditionTrue,
+		Reason:  "IssueIsOpen",
+		Message: "Issue is open",
+	}
+	if state != "open" {
+		condition = &metav1.Condition{
+			Type:    "IssueIsOpen",
+			Status:  metav1.ConditionFalse,
+			Reason:  fmt.Sprintf("IssueIs%s", state),
+			Message: fmt.Sprintf("Issue is %s", state),
+		}
+	}
+	if !meta.IsStatusConditionPresentAndEqual(issueObject.Status.Conditions, "IssueIsOpen", condition.Status) {
+		meta.SetStatusCondition(&issueObject.Status.Conditions, *condition)
+		return true
+	}
+	return false
+}
+
+// CheckForPr checks if the issue has an open PR
+func (r *GithubIssueReconciler) CheckForPr(githubIssue *github.Issue, issueObject *issues.GithubIssue) bool {
+	if githubIssue == nil {
+		return false
+	}
+	condition := &metav1.Condition{
+		Type:    "IssueHasPR",
+		Status:  metav1.ConditionFalse,
+		Reason:  "IssueHasNoPR",
+		Message: "Issue has no PR",
+	}
+	if githubIssue.GetPullRequestLinks() != nil {
+		condition = &metav1.Condition{
+			Type:    "IssueHasPR",
+			Status:  metav1.ConditionTrue,
+			Reason:  "IssueHasPR",
+			Message: "Issue has an open PR",
+		}
+	}
+	if !meta.IsStatusConditionPresentAndEqual(issueObject.Status.Conditions, "IssueHasPR", condition.Status) {
+		meta.SetStatusCondition(&issueObject.Status.Conditions, *condition)
+		return true
+	}
+	return false
+}
+
+// fetchAllIssues gets all issues in a repository with retry for rate limits
+func (r *GithubIssueReconciler) fetchAllIssues(ctx context.Context, owner string, repo string) ([]*github.Issue, error) {
+	opt := &github.IssueListByRepoOptions{}
+	maxRetries := 5
+
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		allIssues, response, err := r.GitHubClient.Issues.ListByRepo(ctx, owner, repo, opt)
+		if err == nil {
+			r.Log.Info("Fetched issues successfully")
+			return allIssues, nil
+		}
+		if response != nil && response.StatusCode == 403 {
+			resetTime := response.Rate.Reset.Time
+			waitDuration := time.Until(resetTime) + time.Second
+			r.Log.Warn(fmt.Sprintf("Rate limit hit. Retrying after %v seconds.", waitDuration.Seconds()))
+			time.Sleep(waitDuration)
+			continue
+		}
+		return nil, fmt.Errorf("error fetching issues: %v", err)
+	}
+	return nil, fmt.Errorf("exceeded retries fetching issues")
+}
+
+// CloseIssue closes the issue on GitHub
+func (r *GithubIssueReconciler) CloseIssue(ctx context.Context, owner string, repo string, gitHubIssue *github.Issue) error {
+	if gitHubIssue == nil {
+		return errors.New("could not find issue in repository")
+	}
+	state := "closed"
+	closedIssueRequest := &github.IssueRequest{State: &state}
+	_, _, err := r.GitHubClient.Issues.Edit(ctx, owner, repo, *gitHubIssue.Number, closedIssueRequest)
+	if err != nil {
+		return fmt.Errorf("failed to close GitHub issue: %v", err)
+	}
```

Comment:
Why are you throwing away the returned status code? It may be possible that no error has returned but the operation did not succeed (i.e. an invalid token)